tests/test-update-direct.c

```diff
@@ -0,0 +1,255 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2018 Endless Mobile, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Joaquim Rocha <jrocha@endlessm.com>
+ */
+
+#include <test-common/misc-utils.h>
+#include <test-common/spawn-utils.h>
+#include <test-common/utils.h>
+
+#include <gio/gio.h>
+#include <locale.h>
+
+#include "eos-updater-generated.h"
+
+typedef enum {
+	EOS_UPDATER_STATE_NONE = 0,
+	EOS_UPDATER_STATE_READY,
+	EOS_UPDATER_STATE_ERROR,
+	EOS_UPDATER_STATE_POLLING,
+	EOS_UPDATER_STATE_UPDATE_AVAILABLE,
+	EOS_UPDATER_STATE_FETCHING,
+	EOS_UPDATER_STATE_UPDATE_READY,
+	EOS_UPDATER_STATE_APPLYING_UPDATE,
+	EOS_UPDATER_STATE_UPDATE_APPLIED,
+        EOS_UPDATER_N_STATES,
+} EosUpdaterState;
+
+gboolean cancelled_states[EOS_UPDATER_N_STATES] = { FALSE };
+const gchar *states_str[] = {"None",
+                             "Ready",
+                             "Error",
+                             "Polling",
+                             "UpdateAvailable",
+                             "Fetching",
+                             "UpdateReady",
+                             "ApplyingUpdate",
+                             "UpdateApplied"
+};
+
+guint cancelled_error_count = 0;
+guint cancel_calls_count = 0;
```

Comment:
Nitpick: I’d prefer to keep the test state all within locals in `test_cancel_update()` rather than as static globals. Means the test is a bit better encapsulated.

Reply:
Alright...